pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller.go

```diff
@@ -0,0 +1,231 @@
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apisinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/apidefinition"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+const (
+	ControllerName = "kcp-virtual-apiexport-api-reconciler"
+	byWorkspace    = ControllerName + "-byWorkspace" // will go away with scoping
+)
+
+type CreateAPIDefinitionFunc func(apiResourceSchema *apisv1alpha1.APIResourceSchema, version string, identityHash string) (apidefinition.APIDefinition, error)
+
+// NewAPIReconciler returns a new controller which reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+func NewAPIReconciler(
+	kcpClusterClient kcpclient.ClusterInterface,
+	apiResourceSchemaInformer apisinformer.APIResourceSchemaInformer,
+	apiExportInformer apisinformer.APIExportInformer,
+	createAPIDefinition CreateAPIDefinitionFunc,
+) (*APIReconciler, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerName)
+
+	c := &APIReconciler{
+		kcpClusterClient: kcpClusterClient,
+
+		apiResourceSchemaLister:  apiResourceSchemaInformer.Lister(),
+		apiResourceSchemaIndexer: apiResourceSchemaInformer.Informer().GetIndexer(),
+
+		apiExportLister:  apiExportInformer.Lister(),
+		apiExportIndexer: apiExportInformer.Informer().GetIndexer(),
+
+		queue: queue,
+
+		createAPIDefinition: createAPIDefinition,
+
+		apiSets: map[dynamiccontext.APIDomainKey]apidefinition.APIDefinitionSet{},
+	}
+
+	if err := apiExportInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace: indexByWorksapce,
+	}); err != nil {
+		return nil, err
+	}
+
+	apiExportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIExport(obj)
+		},
+	})
+
+	apiResourceSchemaInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIResourceSchema(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIResourceSchema(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// APIReconciler is a controller that reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+type APIReconciler struct {
```

Comment:
nit: we don't actually reconcile anything, just collect an in-memory cache, right?